homeassistant/components/screenlogic/data.py

```diff
@@ -0,0 +1,298 @@
+
+
+class ScreenLogicDataRule(ScreenLogicRule):
+    """Represents a data rule."""
+
+    def __init__(
+        self, test: Callable[..., bool], test_path_template: tuple[str | int, ...]
+    ) -> None:
+        """Initialize a ScreenLogic data rule."""
+        self._test_path_template = test_path_template
+        super().__init__(test)
+
+    def test(
+        self, gateway: ScreenLogicGateway, data_path: tuple[str | int, ...]
+    ) -> bool:
+        """Check the rule against the gateway's data."""
+        test_path = realize_path_template(self._test_path_template, data_path)
+        return self._test(gateway.get_data(*test_path))
+
+
+class ScreenLogicEquipmentRule(ScreenLogicRule):
+    """Represents an equipment flag rule."""
+
+    def test(
+        self, gateway: ScreenLogicGateway, data_path: tuple[str | int, ...]
+    ) -> bool:
+        """Check the rule against the gateway's equipment flags."""
+        return self._test(gateway.equipment_flags)
+
+
+TemplateParameters = dict[
+    EntityParameter,
+    Callable | EntityCategory | int | ScreenLogicRule | tuple,
+]
+
+
+TemplateValues = dict[str, TemplateParameters | dict]
+
+TemplateGroups = dict[int | str, TemplateValues]
+
+TemplateDevices = dict[str, TemplateGroups]
+
+TemplateData = dict[str, TemplateGroups]
+
+
+DEVICE_INCLUSION_RULES = {
+    DEVICE.PUMP: ScreenLogicDataRule(
+        lambda pump_data: pump_data[VALUE.DATA] != 0,
+        (PathPart.DEVICE, PathPart.INDEX),
+    ),
+    DEVICE.INTELLICHEM: ScreenLogicEquipmentRule(
+        lambda flags: EQUIPMENT_FLAG.INTELLICHEM in flags,
+    ),
+    DEVICE.SCG: ScreenLogicEquipmentRule(
+        lambda flags: EQUIPMENT_FLAG.CHLORINATOR in flags,
+    ),
+}
+
+DEVICE_SUBSCRIPTION = {
+    DEVICE.CONTROLLER: CODE.STATUS_CHANGED,
+    DEVICE.INTELLICHEM: CODE.CHEMISTRY_CHANGED,
+}
+
+
+def get_ha_unit(entity_data: dict) -> StrEnum | str | None:
+    """Return a Home Assistant unit of measurement from a UNIT."""
+    sl_unit = entity_data.get(ATTR.UNIT)
+    return SL_UNIT_TO_HA_UNIT.get(sl_unit, sl_unit)
+
+
+def realize_path_template(
+    template_path: tuple[str | int, ...], data_path: tuple[str | int, ...]
+) -> tuple[str | int, ...]:
+    """Make data path from template and current."""
+    if not data_path or len(data_path) < 3:
+        raise KeyError(
+            f"Missing or invalid required parameter: 'data_path' for template path '{template_path}'"
+        )
+    device, group, data_key = data_path
+    realized_path = []
+    for part in template_path:
+        match part:
+            case PathPart.DEVICE:
+                realized_path.append(device)
+            case PathPart.INDEX | PathPart.KEY:
+                realized_path.append(group)
+            case PathPart.VALUE:
+                realized_path.append(data_key)
+            case _:
+                realized_path.append(part)
+
+    return tuple(realized_path)
+
+
+def process_entity(gateway: ScreenLogicGateway, template_entities: TemplateData):
```

Comment:
Its a bit confusing to call this template data since its easily confused with jinja2 templates where are so prevalent in the codebase

Reply:
Maybe `categorize_entity`..naming is hard here

Reply:
I can rename this.

Reply:
For clarity, is it the word "template", or that "data" is too generic with "template"? Does `TemplateDevices`, `TemplateGroups`, `TemplateValues`, `TemplateParameters` have the same issue?